### PR TITLE
Create `MultiFunctionTemplateModel` and use in ConvFit

### LINKED ISSUE #36964: Create `MultiFunctionTemplateModel` and use in presenter
**Description**
We have a `MultiFunctionTemplateView` and `MultiFunctionTemplatePresenter` which are almost ready to be used by the IqtFit tab. Currently they are just used by ConvFit.

We want to create a `MultiFunctionTemplateModel` so that it can be used in the `MultiFunctionTemplatePresenter`, and the `ConvFunctionTemplateModel` should inherit from this new model. We will then be able to reuse some code for the `IqtFunctionTemplateModel` when we get to that stage.

**Acceptance Criteria**
- `ConvFunctionTemplateModel` inherits from `MultiFunctionTemplateModel`
- `MultiFunctionTemplatePresenter` uses the new `MultiFunctionTemplateModel`
- Common functionality for `ConvFunctionTemplateModel` and `IqtFunctionTemplateModel` has been moved to `MultiFunctionTemplateModel`

### LINKED ISSUE #37004: Crash in IqtFit for invalid data after setting 1 exponential
**Describe the bug**
There is a crash when after loading invalid data into the IqtFit tab of Data Analysis, and then setting the number of exponentials to 1. This crash should not happen.

**To Reproduce**

1. Produce an Iqt workspace from Data Manipulation but use the irs26176_graphite002_red file in both boxes
2. Load the Iqt workspace into the Data Analysis GUI
3. Change the number of exponentials to 1
4. Crash.

**Expected behavior**
There should be no crash.

**Platform/Version (please complete the following information):**
 - OS: All

### PR DESCRIPTION
**Pull request by robertapplin**
### Description of work
This PR creates a `MultiFunctionTemplateModel` class and gets the `ConvFunctionTemplateModel` and `IqtFunctionTemplateModel` to inherit from it. I have then removed all references of `ConvFunctionTemplateModel` in the `MultiFunctionTemplatePresenter` in favour of `MultiFunctionTemplateModel`.

The next step in a future PR will be to remove `IqtTemplatePresenter` and `IqtTemplateView` and replace them with the already implemented `MultiFunctionTemplatePresenter` and `MultiFunctionTemplateView`.

Fixes #36964

Fixes #37004 

I have sent an unstable windows package to Spencer for testing: https://builds.mantidproject.org/job/build_packages_from_branch/755/

### To test:
Code review

Follow the testing instructions here
https://developer.mantidproject.org/Testing/Inelastic/DataAnalysisTests.html

Also fixes #37004, so follow these instructions:
1. Produce an Iqt workspace from Data Manipulation but use the irs26176_graphite002_red file in both boxes
2. Load the Iqt workspace into the Data Analysis GUI
3. Change the number of exponentials to 1
4. Crash.

*This does not require release notes* because **there are no functional changes**

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: qt/scientific_interfaces/Inelastic/Analysis/FunctionBrowser/MultiFunctionTemplateModel.cpp [hunk 1]
```diff
@@ -226,24 +226,24 @@ void MultiFunctionTemplateModel::setParameter(ParamID name, double value) {
   }
 }
 
-boost::optional<double> MultiFunctionTemplateModel::getParameter(ParamID name) const {
+std::optional<double> MultiFunctionTemplateModel::getParameter(ParamID name) const {
   auto const paramName = getParameterName(name);
-  return paramName ? m_model->getParameter(*paramName) : boost::optional<double>();
+  return paramName ? m_model->getParameter(*paramName) : std::optional<double>();
```
Comment (adriazalvarez) [inline qt/scientific_interfaces/Inelastic/Analysis/FunctionBrowser/MultiFunctionTemplateModel.cpp:232]: Is this initialization `std::optional<double>` equivalent to `std::nullopt`?

File: qt/scientific_interfaces/Inelastic/Analysis/FunctionBrowser/IqtFunctionTemplateModel.cpp [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+
+#include <map>
+#include <tuple>
+
+namespace {
+using namespace MantidQt::CustomInterfaces::IDA;
+
+std::tuple<double, double> calculateLifetimeAndHeight(Mantid::MantidVec const &x, Mantid::MantidVec const &y) {
+  auto lifeTime = (x[1] - x[0]) / (log(y[0]) - log(y[1]));
+  if (lifeTime <= 0 || std::isinf(lifeTime))
```
Comment (adriazalvarez) [inline qt/scientific_interfaces/Inelastic/Analysis/FunctionBrowser/IqtFunctionTemplateModel.cpp:21]: Perhaps we can avoid computing the infinite, 
```suggestion
  auto lifeTime= ((x[0]>=x[1]) || (y[0]==y[1])) ? 1.0 : (x[1] - x[0]) / (log(y[0]) - log(y[1]));
```

**Review by adriazalvarez — CHANGES_REQUESTED** [review]
I have tested changes and it works well. Also the crash is fixed. 
I think is a good refactoring. 
I have two minor questions, and then I'll approve.

File: qt/scientific_interfaces/Inelastic/Analysis/FunctionBrowser/MultiFunctionTemplateModel.cpp (hunk 1 above)
Comment (robertapplin) [reply qt/scientific_interfaces/Inelastic/Analysis/FunctionBrowser/MultiFunctionTemplateModel.cpp:232]: Good question, no I believe `std::optional<double>` instantiates the default value for double, which is 0.0. I think this isn't very clear so I'll change it!

Comment (robertapplin) [reply qt/scientific_interfaces/Inelastic/Analysis/FunctionBrowser/MultiFunctionTemplateModel.cpp:232]: Actually, I think these should probably return nullopt, so good spot

Comment (robertapplin) [reply qt/scientific_interfaces/Inelastic/Analysis/FunctionBrowser/MultiFunctionTemplateModel.cpp]: I changed my mind again, I think some of these functions should explicitly return the default value for either double or std::string, and the `getParameterName` function should return a std::nullopt. This is based on what they are being used for in the rest of the model

File: qt/scientific_interfaces/Inelastic/Analysis/FunctionBrowser/IqtFunctionTemplateModel.cpp (hunk 2 above)
Comment (robertapplin) [reply qt/scientific_interfaces/Inelastic/Analysis/FunctionBrowser/IqtFunctionTemplateModel.cpp]: I've added a check to make sure the denominator is not zero. It is a bit more difficult to check the lifetime is not going to be below zero before doing the calculation, so I have left this in an if statement

**Review by adriazalvarez — DISMISSED** [review]
Thanks for answering my doubts.

**Review by adriazalvarez — APPROVED** [review]

**Pull request merged**